# 4 Data Consistency Checks

Contents

1. Importing libraries
2. Import data
3. Checking Products dataframe
    3.1 Missing values and duplicates
    3.2 Testing for mixed datatypes
    3.3 Correcting error in aisle_id
    3.4 Final data checks
4. Checking Orders dataframe
    4.1 Testing for mixed datatypes
    4.2 Missing values and duplicates
5. Exporting data

1. Importing libraries

In [24]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [25]:
path = r'C:\Users\Caroline\Documents\Career Foundry\03 2023 Instacart Basket Analysis'

2. Import Data

In [26]:
#import orders_wrangled.csv as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [27]:
#import products_wrangled.csv as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_wrangled.csv'), index_col = 0)

3. Checking Products dataframe

3.1 Missing values and duplicates

In [28]:
#  to find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [29]:
# create a dataframe with nulls in the product_name column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [30]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [31]:
# Create a dataframe where all the missing product names are not included 
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [32]:
# check for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [33]:
# Which values are duplicated?
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [34]:
# check the current number of rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [35]:
# new data frome without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [36]:
# check shape of products dataframe after removing duplicates
df_prods_clean_no_dups.shape

(49672, 5)

3.2 Testing for mixed datatypes

In [83]:
# Search for mixed data types in df_prods_clean_no_dups
for col in df_prods_clean_no_dups.columns.tolist():
  weird = (df_prods_clean_no_dups[[col]].applymap(type) != df_prods_clean_no_dups[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods_clean_no_dups[weird]) > 0:
    print (col)
print ('test complete - this statement will always print')

test complete - this statement will always print


3.3 Correcting error in aisle_id

In [37]:
#Descriptive statistics of df_prods_clean_no_dups
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     58.320664      11.728942      9.993282
std    14340.705287     51.109052       5.850779    453.615536
min        1.000000   -128.000000       1.000000      1.000000
25%    12432.750000     28.000000       7.000000      4.100000
50%    24850.500000     65.000000      13.000000      7.100000
75%    37268.250000     98.000000      17.000000     11.100000
max    49688.000000    127.000000      21.000000  99999.000000

A negative value for aisle_id is odd, and the max value in prices is unusual. 

In [38]:
# Search for records with negative aisle_is value
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['aisle_id'] < 0]

product_id                           product_name  aisle_id  \
32             33          Organic Spaghetti Style Pasta      -125   
51             52                    Mirabelle Brut Rose      -122   
57             58                  Whole Wheat Tortillas      -128   
67             68                Pancake Mix, Buttermilk      -126   
82             83            100% Whole Wheat Pita Bread      -128   
...           ...                                    ...       ...   
49652       49648                   Squid Ink Fettuccini      -125   
49660       49656             Masada Kosher Pocket Bread      -128   
49665       49661                                  Porto      -122   
49677       49673  Stone Baked Multi Grain Artisan Rolls      -127   
49682       49678        Large Chicken & Cheese Taquitos      -127   

       department_id  prices  
32                 9    12.6  
51                 5    14.4  
57                 3     6.7  
67                14    13.7  
82                 3    14.4  
...              ...     ...  
49652              9     2.4  
49660              3     7.1  
49665              5     8.2  
49677              1     5.6  
49682              1     3.4  

[1832 rows x 5 columns]

In [39]:
#import original products.csv as df_prods_ori
df_prods_ori = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [40]:
# Search for records with negative aisle_id value
df_prods_ori.loc[df_prods_ori['aisle_id'] < 0]

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

After doing research, changing the data type to int8 in Task 4.4 turned values that exceeded 127 to negative values. Aisle_id had values to 134. Solution - overwrite the aisle_id from original data after repeating missing values & duplicate value steps so that the indexes of both dataframes match.

In [41]:
# Create a dataframe where all the missing product names are not included 
df_prods_clean2 = df_prods_ori[df_prods_ori['product_name'].isnull() == False]

In [42]:
# new data frome without duplicates
df_prods_clean2_no_dups = df_prods_clean2.drop_duplicates()

In [43]:
# check shape of products dataframe after removing duplicates
df_prods_clean2_no_dups.shape

(49672, 5)

In [48]:
#overwrite aisle_id column on df_prods_clean_no_dups
df_prods_clean_no_dups.eval('aisle_id = @df_prods_clean2_no_dups.aisle_id', inplace = True)

3.4 Final data checks

In [49]:
# Re-check the descriptive statistics of df_prods_clean_no_dups
df_prods_clean_no_dups.describe()

product_id  department_id        prices      aisle_id
count  49672.000000   49672.000000  49672.000000  49672.000000
mean   24850.349775      11.728942      9.993282     67.762442
std    14340.705287       5.850779    453.615536     38.315784
min        1.000000       1.000000      1.000000      1.000000
25%    12432.750000       7.000000      4.100000     35.000000
50%    24850.500000      13.000000      7.100000     69.000000
75%    37268.250000      17.000000     11.100000    100.000000
max    49688.000000      21.000000  99999.000000    134.000000

In [51]:
# Search for records with prices over 40
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] > 40]

product_id                      product_name  department_id   prices  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese             16  14900.0   
33666       33664             2 % Reduced Fat  Milk             16  99999.0   

       aisle_id  
21554       108  
33666        84

In [60]:
#  Find median value of column 'prices'
df_prods_clean_no_dups['prices'].median()

7.1

In [61]:
# Replace 'prices' for product_id 21533 and 33666 with the median value 
df_prods_clean_no_dups = df_prods_clean_no_dups.replace ({"prices":{99999.0: 7.1, 14900.0:7.1 }})

In [77]:
# check that changes were sucessful
df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"]==21553]

product_id                      product_name  department_id  prices  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese             16     7.1   

       aisle_id  
21554       108

In [79]:
df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"]==33664]

product_id           product_name  department_id  prices  aisle_id
33666       33664  2 % Reduced Fat  Milk             16     7.1        84

In [63]:
# Re-check the descriptive statistics of df_prods_clean_no_dups
df_prods_clean_no_dups.describe()

product_id  department_id        prices      aisle_id
count  49672.000000   49672.000000  49672.000000  49672.000000
mean   24850.349775      11.728942      7.680414     67.762442
std    14340.705287       5.850779      4.199298     38.315784
min        1.000000       1.000000      1.000000      1.000000
25%    12432.750000       7.000000      4.100000     35.000000
50%    24850.500000      13.000000      7.100000     69.000000
75%    37268.250000      17.000000     11.100000    100.000000
max    49688.000000      21.000000     25.000000    134.000000

Descriptive data looks good. Max price of 25 is reasonable. No negative values.

In [65]:
#Export df_prod_clean_no_dups 
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Checking Orders DF for missing values and duplications 

4. Checking Orders dataframe

In [16]:
# describe the Orders dataframe
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [66]:
# Check df_ords size
df_ords.shape

(3421083, 6)

2. The 'days_since_prior_order_ column has a count fewer then the other column. Possible missing vales that might need to be flagged or dropped.
Min values are all either 1 or 0. Max order_id is the same as the count. Fewer user_id then the count makes sense if users have multiple orders. 0-7 for orders_day_of_the_week for 7 total makes sense. and 0-23 for order_hour_of_day. The maximum value of days_since_prior_order is 30.

2 - Correction - O-6 is the correct range for orders_day_of_the_week. Since the count starts at 0, there are 7 total values.

4.1 Testing for mixed datatypes

In [24]:
# Search for mixed data type in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
print ('test complete - this statement will always print')

test complete - this statement will always print 


#4 If you find mixed-type data, fix it. The column in question should contain observations of a single data type
No changes made.

4.2 Check for missing values and duplicates

In [19]:
# Run a check for missing values in df_ords
#  to find missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#5a There are 206,209 null values in the 'days since prior order' column. This is likely to be customers who have only used Instacart once

In [67]:
#6 Address the missing values using an appropriate method
# Create subset to show missing values in df_ords
df_ords_missing = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [68]:
# View missing values subset
df_ords_missing

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [72]:
# If the max value of 'order_number' is 1, then all of the missing values in days since prior orders are from customers 
# who have only used the servuce once
df_ords_missing['order_number'].max()

1

#6 It does not make sense to impute values in this situation or to remove the data. One option is to add a column to indicate a repeat or non-repeat customer. However a value of 1 in the 'order_number' column can also as a flag for a new customer so I have chosen not to make any changes yet.

In [80]:
# 7. Run a check for duplicate values in your df_ords data
df_dups_ords = df_ords[df_ords.duplicated()]

In [81]:
df_dups_ords

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#7a There are no full duplicates. No changes to be made.

5. Exporting data

In [ ]:
#Export df_prod_clean_no_dups 
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [82]:
#Export df_ords 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))